[profiler] use swap in RangeEventList::consolidate to reduce time the lock is held.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35055 [profiler] Allow record_function ctx manager to profile futures
* **#34499 [profiler] use swap in RangeEventList::consolidate to reduce time the lock is held.**
getEventList.record()

`RangeEventList::consolidate` currently iterates through `blocks`, which
we lock access to and accumulates them into `result`. Instead of doing
this, we can swap with an empty `forward_list` in constant time, and then
unlock, and use this local list in order to populate `result`. This was suggested by @mrshenli 

Differential Revision: [D20346423](https://our.internmc.facebook.com/intern/diff/D20346423/)

rohan-varma committed:
[profiler] use swap in allocBlock to reduce time the lock is held.

RangeEventList::allocBlock currently iterates through `blocks`, which
we serialize access to and accumulates them into `result`. Instead of doing
this, we can swap with an empty `forward_list` in constant time, and then
unlock, and use this local list in order to populate `result`.

Differential Revision: [D20346423](https://our.internmc.facebook.com/intern/diff/D20346423/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler] use swap in allocBlock to reduce time the lock is held."

getEventList.record()

`RangeEventList::allocBlock` currently iterates through `blocks`, which
we lock access to and accumulates them into `result`. Instead of doing
this, we can swap with an empty `forward_list` in constant time, and then
unlock, and use this local list in order to populate `result`.

Differential Revision: [D20346423](https://our.internmc.facebook.com/intern/diff/D20346423/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/autograd/profiler.h:
```diff
@@ -199,14 +199,17 @@ struct RangeEventList {
 
   std::vector<Event> consolidate() {
-    std::lock_guard<std::mutex> guard(mutex_);
+    std::unique_lock<std::mutex> guard(mutex_);
```
this is no longer a guard. Shall we call it lock?

mrshenli reviewed (approved):
LGTM!

rohan-varma committed:
Update on "[profiler] use swap in RangeEventList::consolidate to reduce time the lock is held."

getEventList.record()
getEventList.record()

`RangeEventList::consolidate` currently iterates through `blocks`, which
we lock access to and accumulates them into `result`. Instead of doing
this, we can swap with an empty `forward_list` in constant time, and then
unlock, and use this local list in order to populate `result`. This was suggested by @mrshenli 

Differential Revision: [D20346423](https://our.internmc.facebook.com/intern/diff/D20346423/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler] use swap in RangeEventList::consolidate to reduce time the lock is held."

getEventList.record()

`RangeEventList::consolidate` currently iterates through `blocks`, which
we lock access to and accumulates them into `result`. Instead of doing
this, we can swap with an empty `forward_list` in constant time, and then
unlock, and use this local list in order to populate `result`. This was suggested by @mrshenli 

Differential Revision: [D20346423](https://our.internmc.facebook.com/intern/diff/D20346423/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler] use swap in RangeEventList::consolidate to reduce time the lock is held."

getEventList.record()

`RangeEventList::consolidate` currently iterates through `blocks`, which
we lock access to and accumulates them into `result`. Instead of doing
this, we can swap with an empty `forward_list` in constant time, and then
unlock, and use this local list in order to populate `result`. This was suggested by @mrshenli 

Differential Revision: [D20346423](https://our.internmc.facebook.com/intern/diff/D20346423/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[profiler] use swap in RangeEventList::consolidate to reduce time the lock is held."

getEventList.record()

`RangeEventList::consolidate` currently iterates through `blocks`, which
we lock access to and accumulates them into `result`. Instead of doing
this, we can swap with an empty `forward_list` in constant time, and then
unlock, and use this local list in order to populate `result`. This was suggested by @mrshenli 

Differential Revision: [D20346423](https://our.internmc.facebook.com/intern/diff/D20346423/)

[ghstack-poisoned]
